### ISSUE TITLE
Taggen von Rechnungen und Belegen

### ISSUE THREAD

**Issue by d135-1r43** [hopps-app/hopps#154]
# User Story: Automatische Tag-Zuweisung für Rechnungen

## Beschreibung
Als **Mitarbeiter** möchte ich, dass hochgeladene Rechnungen automatisch mit 4-10 passenden Tags versehen werden, wie z. B. "Baumarkt", "Essen", "Fastfood" oder "Sportartikel", um eine bessere Kategorisierung und Analyse zu ermöglichen.

## Akzeptanzkriterien
1. **Tag-Ermittlung**:
   - Tags werden mithilfe von OpenAI ermittelt.
2. **Eingangsvariable**:
   - Der Service akzeptiert ein beliebiges JSON, das eine Rechnung oder einen Beleg repräsentiert.
   - Es besteht keine Typsicherheit auf der Eingangsvariable, sodass die Rohdaten vom Azure oder Zugferd Service verwendet werden können. 
3. **Ausgabe**:
   - Der Service liefert eine Liste von Tags als Ergebnis.
4. **Architektur**:
   - Der Service wird als eigenständiger Microservice mit dem Namen `fin-narrator` implementiert.
5. **Synchronität**:
   - Die Kommunikation erfolgt über REST, um eine synchrone Anzeige der Tags in der UI zu ermöglichen.
   - Kein Einsatz von Kafka oder asynchroner Kommunikation.

## Zusatzinformationen
- Der Service soll hochperformant und skalierbar sein, da potenziell mehrere Anfragen parallel verarbeitet werden müssen.
- Der Fokus liegt auf der korrekten Erkennung und Zuweisung relevanter Tags.

## Technische Randbedingungen
- Programmiersprache: bevorzugt **Quarkus** für eine schnelle und leichtgewichtige Implementierung.
- Kein Auth nötig, da er nicht nach außen exposed werden wird.

**Comment by schitcrafter** [comment]
> Es besteht keine Typsicherheit auf der Eingangsvariable

> Der Service muss validierte JSON-Daten akzeptieren und auf potenzielle Fehlermeldungen (z. B. fehlerhafte JSON-Struktur) reagieren.

Kannst du das bitte erläutern @d135-1r43? Das Erkennen einer fehlerhaften JSON-Struktur bedeutet ja irgendwie Typsicherheit

**Comment by d135-1r43** [comment]
Korrekt, ich hab es korrigiert.

**Comment by byte-sized-emi** [comment]
@d135-1r43 soll die Liste an möglichen Tags im vornherein festgelegt sein / bei der request angegeben werden, oder wird alles akzeptiert was ChatGPT meint was passend ist?

**Comment by d135-1r43** [comment]
> [@d135-1r43](https://github.com/d135-1r43) soll die Liste an möglichen Tags im vornherein festgelegt sein / bei der request angegeben werden, oder wird alles akzeptiert was ChatGPT meint was passend ist?

Wir akzeptieren alles, was ChatGPT gibt. Achte aber bitte darauf, dass die Tags kleingeschrieben und kurz sind. Entweder durch Prompting oder durch algorithmische Validierung.

**Comment by byte-sized-emi** [comment]
Brauchen wir dafür tests? Es gibt von uns aus nicht viel Logik die getestet werden kann, es ist praktisch nur ein kleiner Service der ChatGPT anspricht.

**Comment by d135-1r43** [comment]
Ja, Tests der REST mit RESTAssured wären wichtig. Da kannst du OpenAI auch gern mocken.

**Comment by byte-sized-emi** [comment]
Das Ticket umfasst jetzt auch den gesamten deployment/integration/UI Teil der User Story

**Comment by manuelhummler** [comment]
@blackwolfk-9 will check the ticket and maybe rerefine it
